test/data/data_tx_test.go

```diff
@@ -0,0 +1,527 @@
+
+package datatxtest
+
+import (
+	"io/ioutil"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/hyperledger-labs/orion-server/pkg/constants"
+	"github.com/hyperledger-labs/orion-server/pkg/server/testutils"
+	"github.com/hyperledger-labs/orion-server/pkg/types"
+	"github.com/hyperledger-labs/orion-server/test/setup"
+	"github.com/stretchr/testify/require"
+)
+
+var baseNodePort uint32
+var basePeerPort uint32
+var portMutex sync.Mutex
+
+func init() {
+	baseNodePort = 11123
+	basePeerPort = 11134
+}
+
+func getPorts(num uint32) (node uint32, peer uint32) {
+	portMutex.Lock()
+	defer portMutex.Unlock()
+
+	node = baseNodePort
+	peer = basePeerPort
+	baseNodePort += num
+	basePeerPort += num
+
+	return
+}
+
+func TestDataTx(t *testing.T) {
+	dir, err := ioutil.TempDir("", "int-test")
+	require.NoError(t, err)
+
+	nPort, pPort := getPorts(1)
+	setupConfig := &setup.Config{
+		NumberOfServers:     1,
+		TestDirAbsolutePath: dir,
+		BDBBinaryPath:       "../../bin/bdb",
+		CmdTimeout:          10 * time.Second,
+		BaseNodePort:        nPort,
+		BasePeerPort:        pPort,
+	}
+	c, err := setup.NewCluster(setupConfig)
+	require.NoError(t, err)
+	defer c.ShutdownAndCleanup()
+
+	require.NoError(t, c.Start())
+	leaderIndex := -1
+	require.Eventually(t, func() bool {
+		leaderIndex = c.AgreedLeader(t, 0)
+		return leaderIndex >= 0
+	}, 30*time.Second, 100*time.Millisecond)
+
+	s := c.Servers[0]
+
+	// creating two databases -- db1 and db2
+	setup.CreateDatabases(t, s, []string{"db1", "db2"})
+
+	// creating two users -- alice and bob
+	aliceCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "alice")
+	bobCert, _ := testutils.LoadTestCrypto(t, c.GetUserCertDir(), "bob")
+	setup.CreateUsers(t, s, []*types.UserWrite{
+		{
+			User: &types.User{
+				Id:          "alice",
+				Certificate: aliceCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+		{
+			User: &types.User{
+				Id:          "bob",
+				Certificate: bobCert.Raw,
+				Privilege: &types.Privilege{
+					DbPermission: map[string]types.Privilege_Access{
+						"db1": types.Privilege_ReadWrite,
+						"db2": types.Privilege_ReadWrite,
+					},
+				},
+			},
+		},
+	})
+
+	t.Run("blind write to a single database", func(t *testing.T) {
+		dataTx := &types.DataTx{
+			MustSignUserIds: []string{
+				"alice",
+			},
+			TxId: uuid.New().String(),
+			DbOperations: []*types.DBOperation{
+				{
+					DbName: "db1",
+					DataWrites: []*types.DataWrite{
+						{
+							Key:   "key1",
+							Value: []byte("key1value1"),
+						},
+						{
+							Key:   "key2",
+							Value: []byte("key2value1"),
+							Acl: &types.AccessControl{
+								ReadWriteUsers: map[string]bool{"alice": true},
+							},
+						},
+					},
+				},
+			},
+		}
+
+		aliceSigner, err := s.Signer("alice")
```

Comment:
You can get the signers for alice and bob above, once.

Reply:
Done